Improved distribution of telemetry bandwidth over sensors (fixes #5785).

pulquero opened the pull request:
Simplified code to have less CPU and flash overhead. Instead of defining a pre-defined sensor order, now  inject the esc sensors at an adjustable rate into the list of other sensors.

(Tested on F4, fport, opentx 2.2.1)

jirif commented:
> @pulquero Please try pulquero/betaflight:esc-telemetry branch. I don't believe there is any missing NULL checks since ESC_SENSOR_COMBINED returns a pre-allocated struct.

Perfect it fixing VFAS, but regarding NULL checks, getEscSensorData() return always NULL when FEATURE_ESC_SENSOR is off (in my case)

```
escSensorData_t *getEscSensorData(uint8_t motorNumber)
{
    if (!feature(FEATURE_ESC_SENSOR)) {
        return NULL;
    }

```
so this is proposed solution:
```
@@ -539,12 +539,14 @@ void processSmartPortTelemetry(smartPortPayload_t *payload, volatile bool *clear
                     *clearToSend = false;
                 }
                 break;
             case FSSP_DATAID_RPM        :
                 escData = getEscSensorData(ESC_SENSOR_COMBINED);
-                smartPortSendPackage(id, escData->rpm);
-                *clearToSend = false;
+                if (escData != NULL) {
+                    smartPortSendPackage(id, escData->rpm);
+                    *clearToSend = false;
+                }
                 break;
             case FSSP_DATAID_RPM1       :
             case FSSP_DATAID_RPM2       :
             case FSSP_DATAID_RPM3       :
             case FSSP_DATAID_RPM4       :
@@ -558,12 +560,14 @@ void processSmartPortTelemetry(smartPortPayload_t *payload, volatile bool *clear
                     *clearToSend = false;
                 }
                 break;
             case FSSP_DATAID_TEMP        :
                 escData = getEscSensorData(ESC_SENSOR_COMBINED);
-                smartPortSendPackage(id, escData->temperature);
-                *clearToSend = false;
+                if (escData != NULL) {
+                    smartPortSendPackage(id, escData->temperature);
+                    *clearToSend = false;
+                }
                 break;
             case FSSP_DATAID_TEMP1      :
             case FSSP_DATAID_TEMP2      :
             case FSSP_DATAID_TEMP3      :
             case FSSP_DATAID_TEMP4      :

```

pulquero commented:
@jirif You've given me an idea of how to make this even better.

pulquero commented:
OK, so I was going to use a linked-list of tables so I could dynamically chain together the relevant tables based on what actual features are active. But, I discovered that the presence of some 'empty' slots seem to be needed. In particular, if I commment out ASPD, which isnt used, then I continually get sensor lost for ACCZ, even though every other sensor is updating nicely. It seems to be something to do with that position in the table. Maybe oversaturating the write queue or some such thing????

jirif commented:
case FSSP_DATAID_ASPD: leave cleartosend as is, try comment  out it in case FSSP_DATAID_ACCZ : it will show problem with queue flushing

pulquero commented:
Not sure I understand, but I added *clearToSend = false to ASPD and tried it. Result, T1, Hdg and AccX are all lost.
Next, tried removing ASPD from table and comment out *clearToSend in ACCZ. Result AccX is lost.

You don't think that I've introduced too much overhead (e.g. my use of % operator) and that has introduced some delay upsetting the timing before timeouts?

pulquero commented:
@jirif Please could you re-test. My rewrite doesn't seem to suffer from that dodgy sensor lost problem.
@wind0r That's a nice idea, though it'll be a bit more ugly due to the number of sensors there are here. Maybe a preprocessor macro could be used to make it look more compact. Anyway, I'll leave that for another PR.

I am now happy with this PR and don't intend to make further changes unless requested - merge at convenience.

jirif commented:
@pulquero all working fine. One more improvement, we do not need to feed VFAS and CURRENT explicitly because its already part of relaxed esc telemetry data.

```
@@ -142,12 +142,14 @@ enum
     FSSP_DATAID_A3         = 0x0900 ,
     FSSP_DATAID_A4         = 0x0910
 };
 
 const uint16_t frSkyDataIdTable[] = {
+#ifndef USE_ESC_SENSOR
     FSSP_DATAID_VFAS      ,
     FSSP_DATAID_CURRENT   ,
+#endif
     FSSP_DATAID_FUEL      ,
     FSSP_DATAID_A4        ,
     FSSP_DATAID_T1        ,
     FSSP_DATAID_T2        ,
     FSSP_DATAID_HEADING   ,

```

pulquero commented:
@jirif Yeah, I figured it was simpler not to worry about the extra VFAS/CURRENT here and there - it just means these key sensors have a slight priority boost. I will address it in my next PR, where I'll do @wind0r suggestion - include only sensors based on runtime config.
